orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/clouddriver/config/CloudDriverConfiguration.java

```diff
@@ -0,0 +1,224 @@
+      this.retrofitLogLevel = retrofitLogLevel;
+      this.spinnakerRequestInterceptor = spinnakerRequestInterceptor;
+      this.cloudDriverConfigurationProperties = cloudDriverConfigurationProperties;
+    }
+
+    <T> T buildWriteableService(Class<T> type) {
+      return buildService(type, cloudDriverConfigurationProperties.getCloudDriverBaseUrl());
+    }
+
+    private <T> T buildService(Class<T> type, String url) {
+      return new RestAdapter.Builder()
+          .setRequestInterceptor(spinnakerRequestInterceptor)
+          .setEndpoint(newFixedEndpoint(url))
+          .setClient(
+              new Ok3Client(
+                  clientProvider.getClient(new DefaultServiceEndpoint("clouddriver", url))))
+          .setLogLevel(retrofitLogLevel)
+          .setLog(new RetrofitSlf4jLog(type))
+          .setConverter(new JacksonConverter(objectMapper))
+          .build()
+          .create(type);
+    }
+
+    private <T> SelectableService buildReadOnlyService(Class<T> type) {
+      List<String> urls =
+          cloudDriverConfigurationProperties.getCloudDriverReadOnlyBaseUrls().stream()
+              .map(CloudDriverConfigurationProperties.BaseUrl::getBaseUrl)
+              .collect(Collectors.toList());
+
+      if (urls.isEmpty()
+          || urls.stream()
+              .allMatch(
+                  url ->
+                      url.equalsIgnoreCase(
+                          cloudDriverConfigurationProperties.getCloudDriverBaseUrl()))) {
+        log.info(
+            "No readonly URLs configured for clouddriver, using default URL {} for {}",
+            cloudDriverConfigurationProperties.getCloudDriverBaseUrl(),
+            type.getSimpleName());
+      }
+
+      List<ServiceSelector> selectors = new ArrayList<>();
+      cloudDriverConfigurationProperties
+          .getCloudDriverReadOnlyBaseUrls()
+          .forEach(
+              url -> {
+                ServiceSelector selector =
+                    new DefaultServiceSelector(
+                        buildService(type, url.getBaseUrl()), url.getPriority(), url.getConfig());
+
+                if (url.getConfig() != null && url.getConfig().get("selectorClass") != null) {
+                  try {
+                    Class<ServiceSelector> selectorClass =
+                        (Class<ServiceSelector>)
+                            Class.forName(url.getConfig().get("selectorClass").toString());
+                    selector =
+                        (ServiceSelector)
+                            selectorClass.getConstructors()[0].newInstance(
+                                selector.getService(), selector.getPriority(), url.getConfig());
+                  } catch (Exception e) {
+                    log.error(
+                        "Failed to create selector for class {}, using default service selector for URL {}",
+                        url.getConfig().get("selectorClass"),
+                        url.getBaseUrl());
+                  }
```

Comment:
Would this exception have bubbled up in the groovy code?  I support adding the logging, but is it really OK to swallow the exception?

Reply:
That's a good catch, thanks. Throws a `GroovyCastException` if an invalid selector is given. I think an `IllegalArgumentException` is ok here?

Reply:
Maybe add a test to be sure?  Are you suggesting to catch `IllegalArgumentException` instead of `Exception` (and perhaps re-throw it)?  The more specific the better I guess....though maybe the logging is overkill and whatever exception we get has enough info in it?  A test would tell us.

Reply:
Sure, currently it would throw various different exceptions depending on what was wrong, so I was thinking just catch the Exception and rethrow an IllegalArgumentException with the content. Though I agree the log is redundant - I tested it with an invalid selector class and Orca just fails to start.

The specific exceptions the new code needs to handle are `InstantiationException`, `IllegalAccessException` and `InvocationTargetException`.